### DISCUSSION TITLE
In FastAPI 0.112.3 and later, deprecated on_event startup events are called twice instead of once

### DISCUSSION THREAD

**Discussion by hhartzer** [fastapi/fastapi#13073] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()

@app.on_event("startup")
def on_startup():
    print("Hello, world!")

(This actually doesn't reproduce the issue. Still working on this.)
```

### Description

`on_startup()` gets called twice on FastAPI 0.112.0 and newer. But not 0.110.X 0.111.X is untested.

I know this is deprecated behavior, but it's a pretty major change.

### Operating System

Linux

### FastAPI Version

0.112.3

### Pydantic Version

2.6.3

### Python Version

3.11

**Comment by hhartzer** [comment]
The minimal example does not appear to be working for me. I am guessing this is more complicated than I originally thought.

Still behaving properly. But readily reproducible in a much more complex codebase.

```
(.venv) $ cat app.py
from fastapi import FastAPI, APIRouter

app = FastAPI()
router = APIRouter()
router2 = APIRouter()
app.include_router(router)
app.include_router(router2)

@router.on_event("startup")
def on_startup():
    print("Hello, world!")


@router2.on_event("startup")
async def on_startup2():
    print("Hello, world 2!")
(.venv) $ uvicorn app:app
INFO:     Started server process [23384]
INFO:     Waiting for application startup.
Hello, world!
Hello, world 2!
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
^CINFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [23384]
```

**Reply by hhartzer** [reply]
Narrowing this down further, the issue is fine on 0.112.2 and earlier, and present on 0.112.3 and later.

**Reply by YuriiMotov** [reply]
I think it now works as expected, doesn't it?
You specified two "startup" events and each of them was called once.

It probably worked wrongly in previous versions, but were fixed in the ~0.112-0.113

**Reply by hhartzer** [reply]
No, unfortunately it doesn't. The same function was being called twice. I do have multiple startup event functions, but they are meant to be called once.

There's nothing about this mentioned in the changelog between 0.112.2 and 0.112.3. I don't think the behavior change was intentional.

**Reply by hhartzer** [reply]
Actually, I am mistaken. Seems like this was between 0.112.1 and 0.112.2.

**Comment by YuriiMotov** [comment]
Closing this. Feel free to reopen it when you are ready to provide [MRE](https://stackoverflow.com/help/minimal-reproducible-example)
